lib/ca_certs_stubs.c

```diff
@@ -0,0 +1,51 @@
+#include "caml/alloc.h"
+#include "caml/fail.h"
+#include "caml/memory.h"
+
+#ifdef _WIN32
+
+#include <windows.h>
+
+static value alloc_cons(value v_hd, value v_tl)
+{
+    CAMLparam2(v_hd, v_tl);
+
+    CAMLlocal1(v_cons);
+    v_cons = caml_alloc(2, 1);
+    Store_field(v_cons, 0, v_hd);
+    Store_field(v_cons, 1, v_tl);
+
+    CAMLreturn(v_cons);
+}
```

Comment:
This kind of modularisation has a non-trivial cost in C bindings - it opens a series of roots. This kind of construct is better done using the pre-processor but, to be honest, better not done at all, given the minefield that correctly interfacing with the GC is.

Reply:
In MirageOS, we usually prefer to never allocate _caml_ objects on the C side. On this specific PR, a possible idea should be to pass a callback to the C function which pass values (`string`) to the OCaml side, such as:
```ocaml
external iter_on_anchors : (string -> unit) -> unit = "caml_iter_on_anchors"

let anchors = let res = ref [] in iter_on_anchors (fun anchor -> res := anchor :: !res) in !res
```

The C function still allocate `string`s but it less error-prone than a _complex_ structure like a list.

Reply:
good idea, I just did that. I thought lists were easy enough to do that in C but then I used the wrong tag for the cons cells 🤦‍♂️

Reply:
I know MirageOS does all things for very good reasons, but just noting with an OCaml Dev Team hat that it's incredibly slower to allocate that way! 🙂

Reply:
> it's incredibly slower to allocate that way!

IMHO this should not matter here: The code should be executed only once at binary startup, and is expected to have a small three-digit-number of certificates.